File: javaapp1/Game.java
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

/** 推箱子核心逻辑（Java 教学）。 */
public final class Game {
    public static final class Pos {
        public final int x, y;
        public Pos(int x, int y) { this.x = x; this.y = y; }
        public String key() { return x + "," + y; }
        public Pos off(int dx, int dy) { return new Pos(x + dx, y + dy); }
    }

    private static final class Hist {
        final Pos player;
        final String boxFrom, boxTo;
        Hist(Pos player, String boxFrom, String boxTo) {
            this.player = player;
            this.boxFrom = boxFrom;
            this.boxTo = boxTo;
        }
    }

    public final Set<String> walls = new HashSet<>();
    public final Set<String> goals = new HashSet<>();
    public final Set<String> boxes = new HashSet<>();
    public Pos player = new Pos(0, 0);
    public int moves = 0;
    public boolean won = false;
    public int width = 0;
    public int height = 0;
    public int levelIndex = 0;
    private final List<Hist> hist = new ArrayList<>();

    public static Game fromRows(String[] rows, int index) {
        Game g = new Game();
        g.levelIndex = index;
        int maxX = 0, maxY = 0;
        for (int y = 0; y < rows.length; y++) {
            maxY = y;
            String row = rows[y];
            for (int x = 0; x < row.length(); x++) {
                if (x > maxX) maxX = x;
                char ch = row.charAt(x);
                String k = x + "," + y;
                switch (ch) {
                    case '#': g.walls.add(k); break;
                    case '.': g.goals.add(k); break;
                    case '$': g.boxes.add(k); break;
                    case '*': g.boxes.add(k); g.goals.add(k); break;
                    case '@': g.player = new Pos(x, y); break;
                    case '+': g.player = new Pos(x, y); g.goals.add(k); break;
                    default: break;
                }
            }
        }
        g.width = maxX + 1;
        g.height = maxY + 1;
        return g;
    }

    public boolean tryMove(int dx, int dy) {
        if (won) return false;
        Pos n = player.off(dx, dy);
        String nk = n.key();
        if (walls.contains(nk)) return false;
        if (boxes.contains(nk)) {
            Pos b = n.off(dx, dy);
            String bk = b.key();
            if (walls.contains(bk) || boxes.contains(bk)) return false;
            hist.add(new Hist(player, nk, bk));
            boxes.remove(nk);
            boxes.add(bk);
            player = n;
            moves++;
            checkWin();
            return true;
        }
        hist.add(new Hist(player, null, null));
        player = n;
        return true;
    }

    public boolean undo() {
        if (won || hist.isEmpty()) return false;
        Hist e = null;
        while (!hist.isEmpty()) {
            e = hist.remove(hist.size() - 1);
            if (e.boxFrom != null) break;
            player = e.player;
        }
        if (e == null || e.boxFrom == null) return true;
        player = e.player;
        boxes.remove(e.boxTo);
        boxes.add(e.boxFrom);
        if (moves > 0) moves--;
        won = false;
        return true;
    }

    private void checkWin() {
        for (String b : boxes) {
            if (!goals.contains(b)) {
                won = false;
                return;
            }
        }
        won = true;
    }

    public String renderAscii() {
        StringBuilder sb = new StringBuilder();
        for (int y = 0; y < height; y++) {
            for (int x = 0; x < width; x++) {
                String k = x + "," + y;
                if (player.x == x && player.y == y) {
                    sb.append(goals.contains(k) ? '+' : '@');
                } else if (boxes.contains(k)) {
                    sb.append(goals.contains(k) ? '*' : '$');
                } else if (walls.contains(k)) {
                    sb.append('#');
                } else if (goals.contains(k)) {
                    sb.append('.');
                } else {
                    sb.append(' ');
                }
            }
            sb.append('\n');
        }
        return sb.toString();
    }
}

File: scalaapp1/Game.scala
// 推箱子核心逻辑（Scala 教学）
import scala.collection.mutable

case class Hist(px: Int, py: Int, boxFrom: Option[String] = None, boxTo: Option[String] = None)

class GameState {
  val walls = mutable.Set.empty[String]
  val goals = mutable.Set.empty[String]
  val boxes = mutable.Set.empty[String]
  var px = 0
  var py = 0
  var moves = 0
  var won = false
  var width = 0
  var height = 0
  val hist = mutable.ArrayBuffer.empty[Hist]

  def checkWin(): Unit =
    won = boxes.forall(goals.contains)

  def tryMove(dx: Int, dy: Int): Boolean = {
    if (won) return false
    val nx = px + dx
    val ny = py + dy
    val nk = GameState.key(nx, ny)
    if (walls.contains(nk)) return false
    if (boxes.contains(nk)) {
      val bx = nx + dx
      val by = ny + dy
      val bk = GameState.key(bx, by)
      if (walls.contains(bk) || boxes.contains(bk)) return false
      hist += Hist(px, py, Some(nk), Some(bk))
      boxes -= nk
      boxes += bk
      px = nx; py = ny
      moves += 1
      checkWin()
      true
    } else {
      hist += Hist(px, py)
      px = nx; py = ny
      true
    }
  }

  def undo(): Boolean = {
    if (won || hist.isEmpty) return false
    var entry: Hist = null
    while (hist.nonEmpty) {
      entry = hist.remove(hist.length - 1)
      if (entry.boxFrom.isDefined) {
        px = entry.px; py = entry.py
        boxes -= entry.boxTo.get
        boxes += entry.boxFrom.get
        if (moves > 0) moves -= 1
        won = false
        return true
      }
      px = entry.px; py = entry.py
    }
    true
  }

  def renderAscii(): String = {
    val sb = new StringBuilder
    for (y <- 0 until height; x <- 0 until width) {
      val k = GameState.key(x, y)
      if (px == x && py == y) sb += (if (goals.contains(k)) '+' else '@')
      else if (boxes.contains(k)) sb += (if (goals.contains(k)) '*' else '$')
      else if (walls.contains(k)) sb += '#'
      else if (goals.contains(k)) sb += '.'
      else sb += ' '
      if (x == width - 1) sb += '\n'
    }
    sb.toString
  }
}

object GameState {
  def key(x: Int, y: Int): String = s"$x,$y"

  def fromRows(rows: Seq[String], index: Int = 0): GameState = {
    val s = new GameState
    var maxX = 0
    var maxY = 0
    for ((row, y) <- rows.zipWithIndex) {
      maxY = y
      for ((ch, x) <- row.zipWithIndex) {
        if (x > maxX) maxX = x
        val k = key(x, y)
        ch match {
          case '#' => s.walls += k
          case '.' => s.goals += k
          case '$' => s.boxes += k
          case '*' => s.boxes += k; s.goals += k
          case '@' => s.px = x; s.py = y
          case '+' => s.px = x; s.py = y; s.goals += k
          case _ =>
        }
      }
    }
    s.width = maxX + 1
    s.height = maxY + 1
    s
  }
}
